tech/firewall-rule-format.md

```diff
@@ -575,15 +600,487 @@ For the record, the reason why the protocol is `*` in Azure is that
+      "id": "6979850877904078814",
+      "creationTimestamp": "2018-12-04T04:55:13.187-08:00",
+      "name": "default-allow-ssh",
+      "description": "Allow SSH from anywhere",
+      "network": "https://www.googleapis.com/compute/v1/projects/strong-augury-224506/global/networks/default",
+      "priority": 65534,
+      "sourceRanges": [
+        "0.0.0.0/0"
+      ],
+      "allowed": [
+        {
+          "IPProtocol": "tcp",
+          "ports": [
+            "22"
+          ]
+        }
+      ],
+      "direction": "INGRESS",
+      "logConfig": {
+        "enable": false
+      },
+      "disabled": false,
+      "selfLink": "https://www.googleapis.com/compute/v1/projects/strong-augury-224506/global/firewalls/default-allow-ssh"
+    }
+
+Here is a how a custom firewall rule to allow connectivity to TCP port
+8000 with *Source IP ranges* as `0.0.0.0/0` looks like:
+
+    {
+      "kind": "compute#firewall",
+      "id": "4387657442666205482",
+      "creationTimestamp": "2019-03-26T11:30:29.863-07:00",
+      "name": "allow-8000",
+      "description": "",
+      "network": "https://www.googleapis.com/compute/v1/projects/strong-augury-224506/global/networks/default",
+      "priority": 1000,
+      "sourceRanges": [
+        "0.0.0.0/0"
+      ],
+      "allowed": [
+        {
+          "IPProtocol": "tcp",
+          "ports": [
+            "8000"
+          ]
+        }
+      ],
+      "direction": "INGRESS",
+      "logConfig": {
+        "enable": false
+      },
+      "disabled": false,
+      "selfLink": "https://www.googleapis.com/compute/v1/projects/strong-augury-224506/global/firewalls/allow-8000"
+    }
+
+
+### Any Source: Conclusion
+
+Once we have identified that a critical port such as `22` or `3389` is
+exposed as discussed in the previous section, we need to identify if it
+is exposed to the entire Internet.
+
+In this section, we learn that checking this is not straightforward in
+case of Azure. There are multiple formats to consider in Azure:
+
+  - `"source_address_prefix": "*"`
+  - `"source_address_prefix": "0.0.0.0/0"`
+  - `"source_address_prefix": "Internet"`
+  - `"source_address_prefixes"` list contains `0.0.0.0/0`
+
+With GCP, it is straightforward due to its simple UI and simple JSON
+structure. We only need to check if `0.0.0.0/0` exists in
+`sourceRanges`.
+
+One way to simplify this in Azure could be to normalize the source
+ranges in `source_address_prefix` and `source_address_prefixes` to a
+common format, e.g., normalize both `*` and `Internet` to `0.0.0.0/0`
+and let the source ranges in the common (`com`) bucket be always a list
+even if a single address in `source_address_prefix` is available.
+
+
+Any Port
+--------
+
+### Any Port: Azure
+
+In this section, we will see a couple of ways we can allow connections
+to all ports.
+
+
+#### `destination_port_range` set to `*`
+
+Here is an example where an inbound security rule is created with
+*Destination port ranges* set to `*`.
+
```

Comment:
`Destination port range` is "*". `Destination port ranges` is "[]".

Reply:
The italicized text are labels I have picked up verbatim from Azure Portal UI. So here I mean that I entered `*` in the *Destination port ranges* in the Azure Portal UI.

<img width="585" alt="Screen Shot 2019-04-01 at 2 44 02 PM" src="https://user-images.githubusercontent.com/229488/55316772-bdeb9b80-548c-11e9-8ce2-798a97e78faf.png">

Azure decides to put the port range in either `destination_port_range` or `destination_port_ranges` in the API response depending on whether one or more than one port ranges are specified.

Does this make sense now?

I will clarify this notation convention somewhere at the top of the document and update this pull request.

Reply:
Clarified the notation at [line 46](https://github.com/cloudmarker/guides/pull/5/files#diff-112a678346155cbc8647b65f47882715R46).

Reply:
This makes sense. Thank you for the clarification.